src/algorithms/dqns/qr_dqn.jl

```diff
@@ -0,0 +1,158 @@
+- `target_approximator`::[`AbstractApproximator`](@ref): similar to `approximator`, but used to estimate the target (the next state).
+- `loss_func`: the loss function.
+- `γ::Float32=0.99f0`: discount rate.
+- `batch_size::Int=32`
+- `update_horizon::Int=1`: length of update ('n' in n-step update).
+- `min_replay_history::Int=32`: number of transitions that should be experienced before updating the `approximator`.
+- `update_freq::Int=4`: the frequency of updating the `approximator`.
+- `ensemble_num::Int=1`: the number of ensemble approximators.
+- `target_update_freq::Int=100`: the frequency of syncing `target_approximator`.
+- `stack_size::Union{Int, Nothing}=4`: use the recent `stack_size` frames to form a stacked state.
+- `traces = SARTS`, set to `SLARTSL` if you are to apply to an environment of `FULL_ACTION_SET`.
+- `κ = Float32`, κ in quantile_huber_loss set to 1.0f0.
+- `rng = Random.GLOBAL_RNG`
+"""
+function QRDQNLearner(;
+    approximator::Tq,
+    target_approximator::Tt,
+    loss_func::Tf,
+    stack_size::Union{Int,Nothing} = nothing,
+    γ::Float32 = 0.99f0,
+    batch_size::Int = 32,
+    update_horizon::Int = 1,
+    min_replay_history::Int = 32,
+    update_freq::Int = 1,
+    ensemble_num::Int = 1,
+    target_update_freq::Int = 100,
+    traces = SARTS,
+    update_step = 0,
+    κ::Float32 = 1.0f0,
+    rng = Random.GLOBAL_RNG,
+) where {Tq,Tt,Tf}
+    copyto!(approximator, target_approximator)
+    sampler = NStepBatchSampler{traces}(;
+        γ = γ,
+        n = update_horizon,
+        stack_size = stack_size,
+        batch_size = batch_size,
+    )
+    N = ensemble_num
+    τ = collect(0:N+1)/N
+    τ  = (τ[2:N+1]+τ[1:N])/2
+    QRDQNLearner(
+        approximator,
+        target_approximator,
+        loss_func,
+        min_replay_history,
+        update_freq,
+        update_step,
+        target_update_freq,
+        sampler,
+        N,
+        rng,
+        κ,
+        τ,
+        0.0f0,
+    )
+end
+
+Flux.functor(x::QRDQNLearner) = (Q = x.approximator, Qₜ = x.target_approximator),
+y -> begin
+    x = @set x.approximator = y.Q
+    x = @set x.target_approximator = y.Qₜ
+    x
+end
+
+function (learner::QRDQNLearner)(env)
+    s = send_to_device(device(learner.approximator), state(env))
+    s = Flux.unsqueeze(s, ndims(s) + 1)
+    q = reshape(learner.approximator(s), :, learner.ensemble_num)
+    vec(mean(q, dims = 2)) |> send_to_host
+end
+
+function RLBase.update!(learner::QRDQNLearner, batch::NamedTuple)
+    Q = learner.approximator
+    Qₜ = learner.target_approximator
+    γ = learner.sampler.γ
+    loss_func = learner.loss_func
+    n = learner.sampler.n
+    batch_size = learner.sampler.batch_size
+    N = learner.ensemble_num
+    κ = learner.κ
+    τ = learner.τ
+    D = device(Q)
+
+    s, a, r, t, s′ = (send_to_device(D, batch[x]) for x in SARTS)
+    a = CartesianIndex.(repeat(batch.action, inner = N), 1:(N*batch_size))
+
+    target_q = reshape(Qₜ(s′), :, N, batch_size)
```

Comment:
So here the return of `Qₜ(s′)` is expected to be of size `(n_action, n_quantile, batch_size)` ?

Reply:
Yes, originally it is of size (n_action*n_quantile, batch_size) and it is reshaped to (n_action, n_quantile, batch_size) for convenience.

Reply:
Since Julia is column-major, better to design it as `(n_quantile, n_action, batch_size)` for performance and easier review.

Reply:
Makes sense, will do that.

Reply:
Yes, there is a significant reduction in training time.